### PR TITLE
Add ALL_ATTENTION_FUNCTIONS compatibility for Pixtral model

### PR DESCRIPTION
**Pull request by uminaty**
### What does this PR do?

This PR adds support for `ALL_ATTENTION_FUNCTIONS` to the Pixtral model’s attention mechanism. I added and verified compatibility with `sdpa`, `flash_attention_2`, and `flex_attention`. Since Pixtral also serves as the vision tower in `Mistral 3.1`, users can now set the entire model to use `flash_attention_2`.

I tried to follow the implementation pattern of other models using this interface. For `flash_attention_2`, I reused `position_ids` because the existing attention mask shape isn’t supported. Since Pixtral uses sequence packing and already generates `position_ids`, we leverage [`prepare_fa2_from_position_ids`](https://github.com/huggingface/transformers/blob/afbc293e2b8523bfc92232d8a8d50e29bdce9f41/src/transformers/modeling_flash_attention_utils.py#L184) instead of a mask.

I tested these changes in training and inference: losses match very closely and we observe a 10–25 % throughput improvement depending on the setup.

### Who can review?
- @amyeroberts  
- @qubvel  
- @ArthurZucker

### CODE REVIEW

**Review by qubvel — COMMENTED** [review]
@uminaty thanks for the PR! please use library-defined functions as much as possible 🤗  Thank you!

File: src/transformers/models/pixtral/modeling_pixtral.py [hunk 1]
```diff
@@ -469,6 +528,11 @@ def forward(
             pixel_values: tensor of token features for
                 all tokens of all images of shape (N_toks, D)
         """
+        if image_sizes is None:
+            batch_size, _, height, width = pixel_values.shape
+            # on crée une liste de tuples (H, W) répétée batch_size fois
+            image_sizes = [(height, width)] * batch_size
```
[the remark this answers is not in the document]
Comment (uminaty) [reply src/transformers/models/pixtral/modeling_pixtral.py]: This line is self-explanatory anyway, I think, and I've deleted the comment.

File: src/transformers/models/pixtral/modeling_pixtral.py [hunk 2]
```diff
@@ -172,17 +204,36 @@ def forward(
...
-        attn_weights = nn.functional.dropout(attn_weights, p=self.dropout, training=self.training)
-        attn_output = torch.matmul(attn_weights, value_states)
+        impl = getattr(self.config, "_attn_implementation", "eager")
+
+        if impl == "sdpa" and output_attentions:
+            logger.warning_once(
+                "`torch.nn.functional.scaled_dot_product_attention` does not support `output_attentions=True`. Falling back to "
+                'eager attention. This warning can be removed using the argument `attn_implementation="eager"` when loading the model.'
+            )
+            impl = "eager"
```
[the remark this answers is not in the document]
Comment (uminaty) [reply src/transformers/models/pixtral/modeling_pixtral.py]: Thanks, I've reused the one from multiple files now.

File: src/transformers/models/pixtral/modeling_pixtral.py [hunk 3]
```diff
@@ -132,8 +131,37 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
...
+    scores = torch.matmul(q, k.transpose(-2, -1)) * scaling
+    if attn_mask is not None:
+        scores = scores + attn_mask
+
+    attn = torch.softmax(scores, dim=-1, dtype=torch.float32).to(q.dtype)
+    attn = nn.functional.dropout(attn, p=dropout, training=module.training)
+
+    # (B, H, L, L) @ (B, H, L, D) ➜ (B, H, L, D)
+    out = torch.matmul(attn, v).transpose(1, 2).contiguous()  # (B, L, H, D)
+    return out, attn
```
[the remark this answers is not in the document]
Comment (uminaty) [reply src/transformers/models/pixtral/modeling_pixtral.py]: Yes! I found a similar function used in smolVLM. I copied it.

**Comment by uminaty** [comment]
Thanks @qubvel for the review 🙏! I made the change you suggested, let me know if anything else is needed.

File: src/transformers/models/pixtral/modeling_pixtral.py [hunk 4]
```diff
@@ -172,17 +201,35 @@ def forward(
...
+                    'eager attention. This warning can be removed using the argument `attn_implementation="eager"` when loading the model.'
+                )
+            else:
+                attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
+
+        # Since we use packing, if Flash-Attn 2 is selected we rely on position_ids
+        if self.config._attn_implementation == "flash_attention_2":
+            position_ids = position_ids.to(hidden_states.device, non_blocking=True)
+            attention_mask = None
+            flash_kwargs["position_ids"] = position_ids
```
Comment (ArthurZucker) [inline src/transformers/models/pixtral/modeling_pixtral.py:218]: I would rather we pass this directly as flash_kwargs, and this can be done inside the call of flash attention 2!

**Review by ArthurZucker — APPROVED** [review]
Great addition thanks! We don't really need the position ids, (should be kwargs imo!)

File: src/transformers/models/pixtral/modeling_pixtral.py [hunk 5]
```diff
@@ -381,6 +433,9 @@ class PixtralPreTrainedModel(PreTrainedModel):
     main_input_name = "pixel_values"
     supports_gradient_checkpointing = True
     _no_split_modules = ["PixtralAttentionLayer"]
+    _supports_flash_attn_2 = True
+    _supports_sdpa = True
+    _supports_flex_attn = True
```
Comment (zucchini-nlp) [inline src/transformers/models/pixtral/modeling_pixtral.py:438]: `supports_attention_backends=True` as well

File: src/transformers/models/pixtral/modeling_pixtral.py [hunk 6]
```diff
@@ -494,6 +553,7 @@ def forward(
             patch_embeds,
             attention_mask=attention_mask,
             position_embeddings=position_embeddings,
+            position_ids=position_ids,
```
Comment (zucchini-nlp) [inline src/transformers/models/pixtral/modeling_pixtral.py:556]: we also need to pass `kwargs` here so it can be propagated to each attention layer

**Review by zucchini-nlp — COMMENTED** [review]
Oh cool, I also had a PR for attention in VLMs in https://github.com/huggingface/transformers/pull/37576 😄

File: src/transformers/models/pixtral/modeling_pixtral.py (hunk 5 above)
Comment (uminaty) [reply src/transformers/models/pixtral/modeling_pixtral.py:438]: Thanks! I added it

File: src/transformers/models/pixtral/modeling_pixtral.py (hunk 6 above)
Comment (uminaty) [reply src/transformers/models/pixtral/modeling_pixtral.py:557]: Are you referring to flash_kwargs as in [this comment ](https://github.com/huggingface/transformers/pull/37960#discussion_r2075019859)? Or do you mean passing on the kwargs that are passed (and currently unused) in the `PixtralVisionModel` forward?

Comment (zucchini-nlp) [reply src/transformers/models/pixtral/modeling_pixtral.py:557]: Yeah, the second one. Users should be able to call the models with attention related kwargs directly, and it's passed further downstream

Comment (uminaty) [reply src/transformers/models/pixtral/modeling_pixtral.py]: It's done! And I also used `**kwargs: Unpack[FlashAttentionKwargs],` as in multiple other file I found. Don't hesitate to tell me if my last commit is not what you had in mind.

Comment (zucchini-nlp) [reply src/transformers/models/pixtral/modeling_pixtral.py]: Great, thanks for the `Unpack[FlashAttentionKwargs]`

File: src/transformers/models/pixtral/modeling_pixtral.py (hunk 4 above)
Comment (uminaty) [reply src/transformers/models/pixtral/modeling_pixtral.py]: As @zucchini-nlp suggested, it's good to have kwargs passed through modules anyway, so I used it to pass `position_ids`. Thank you again for the review

**Comment by uminaty** [comment]
Thanks everyone for your reviews! Let me know if anything else is needed before merging 😊

**Comment by ArthurZucker** [comment]
Thanks for the contrib!
